### PR TITLE
HPCC-23062 Compilation fails when cache file cannot be updated

### PR DESCRIPTION
**Pull request by shamser**
Signed-off-by: Shamser Ahmed <shamser.ahmed@lexisnexis.co.uk>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [x] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

### CODE REVIEW

File: ecl/hql/hqlexpr.cpp [hunk 1]
```diff
@@ -1112,7 +1112,15 @@ bool HqlParseContext::createCache(const char * simplifiedEcl, bool isMacro)
         writeStringToStream(*stream, "</Cache>\n");
         stream->flush();
     }
-    cacheFile->move(filename);
+    try
+    {
+        cacheFile->rename(filename);
```
Comment (shamser) [inline ecl/hql/hqlexpr.cpp]: File move is suitable for renames across file systems and always retries on failures.  Rename designed to only works within same file system and does not attempt retries (so, it's quicker).  

Rename is more suitable here as the source & target file is in same directory and retries are unnecessary.

File: ecl/hql/hqlexpr.cpp [hunk 2]
```diff
@@ -1112,7 +1112,15 @@ bool HqlParseContext::createCache(const char * simplifiedEcl, bool isMacro)
...
+    }
+    catch(...)
+    {
+        // compilation shouldn't fail just because the cache couldn't be copied over
```
Comment (richardkchapman) [inline ecl/hql/hqlexpr.cpp]: You'll leak an exception here...

Comment (shamser) [reply ecl/hql/hqlexpr.cpp]: Fixed.

**Comment by shamser** [comment]
@richardkchapman Please can you review.

**Comment by richardkchapman** [comment]
Change looks fine, but we are not generally targeting 7.2.x for any but the most serious of issues (which this does not feel like it is). Can you retarget to 7.4.x?

**Comment by shamser** [comment]
The issue was reported for 7.4.28 .  I'll re-target to 7.4.x.

**Pull request merged**